RCORE-2015 Merge rules for Clear instruction

danieltabacaru opened the pull request:
## What, How & Why?
Given the Clear instruction now has the collection type information, some adjustments have been made to the OT rules:

1. Clear vs Update (same type) => no conflict
2. Clear vs Update (different types) => the tiebreaker is used
3. Clear vs AddInteger => Clear wins


## ☑️ ToDos
* ~~[ ] 📝 Changelog update~~
* [X] 🚦 Tests (or not relevant)
* ~~[ ] C-API, if public C++ API changed~~
* ~~[ ] `bindgen/spec.yml`, if public C++ API changed~~

michael-wb commented on src/realm/sync/transform.cpp:
```diff
@@ -1695,14 +1695,32 @@ DEFINE_MERGE(Instruction::ArrayErase, Instruction::Update)
+        // If both sides are setting/operating on the same type, let them both pass through.
+        // It is important that the instruction application rules reflect this.
+        // If it is not two lists or dictionaries, then the normal last-writer-wins rules will take effect below.
+        if (left.collection_type == CollectionType::List && right.value.type == Type::List) {
+            return;
+        }
+        if (left.collection_type == CollectionType::Dictionary && right.value.type == Type::Dictionary) {
+            return;
+        }
+
```
Should this also include `CollectionType::Set`?

danieltabacaru replied:
No. The client and server disallow creating sets in Mixed properties, so there is no sentinel instruction for a Set.

michael-wb replied:
Do we have any tests that validate this? I didn't see any in the `test_transform_collections_mixed.cpp` file.

jbreams replied:
I don't think we have a test that explicitly tests that you can't set a mixed to a Set (maybe @jedelbo has one I just can't find), but it is validated in `Obj::set_collection()`

danieltabacaru commented on src/realm/sync/instruction_applier.cpp:
```diff
@@ -877,13 +877,30 @@ void InstructionApplier::operator()(const Instruction::ArrayErase& instr)
         }
         void on_list(LstBase& list) override
         {
+            // list property
```
I can add some extra validation since it should always match

jbreams commented on test/test_transform_collections_mixed.cpp:
```diff
@@ -1455,3 +1461,261 @@ TEST(Transform_ClearDictionaryInsideDictionaryVsCreateDictionary)
+    do {
+        // Baseline: property 'any' is not set to any type
+        // {id: 1, any: null}
+        TransformTestHarness h(test_context, {}, [](Obj, ColKey) {});
```
nit, can we initialize the harness after we check if the timestamps are valid so we don't waste time spinning up a server/clients for cases we aren't going to test?

jbreams commented on test/test_transform_collections_mixed.cpp:
```diff
@@ -1455,3 +1461,261 @@ TEST(Transform_ClearDictionaryInsideDictionaryVsCreateDictionary)
+            },
+            t4);
+
+        h.check_merge_result([&](Obj, ColKey) {});
```
Is all we're checking here that the merge doesn't cause a bad changeset?

danieltabacaru replied:
yeah, and that the clients converge

jbreams replied:
Got it. Can you just add a comment saying that these tests are just to make sure that we converge? An empty lambda here makes it seem like these tests are noops.

jbreams commented on test/test_transform_collections_mixed.cpp:
```diff
@@ -1455,3 +1461,261 @@ TEST(Transform_ClearDictionaryInsideDictionaryVsCreateDictionary)
+    do {
+        // Baseline: property 'any' is not set to any type
+        // {id: 1, any: null}
+        TransformTestHarness h(test_context, {}, [](Obj, ColKey) {});
```
Same nit about initializing the harness after we know we're going to use it.

jbreams commented on test/test_transform_collections_mixed.cpp:
```diff
@@ -1455,3 +1461,261 @@ TEST(Transform_ClearDictionaryInsideDictionaryVsCreateDictionary)
+            },
+            t4);
+
+        h.check_merge_result([&](Obj, ColKey) {});
```
Same q about what we're actually testing here.

jbreams replied:
If you don't think that'd be too much work, I think that would be valuable.

danieltabacaru commented:
> Do we know what is causing this error? It seems to be happening (3 times) in this PR, but I'm not sure if it's related to these changes...
> 
> ```
> [2024/04/18 08:38:51.641] 4: sync: client reset
> [2024/04/18 08:38:51.641] 4:   discard local
> [2024/04/18 08:38:51.641] 4:   modify
> [2024/04/18 08:38:51.641] 4:   a Realm can be reset twice
> . . .
> [2024/04/18 08:38:51.641] 4: Assertion failure: /System/Volumes/Data/data/mci/499b8de9879bd994ba18c4339829812e/realm-core/test/object-store/sync/client_reset.cpp:875
> [2024/04/18 08:38:51.641] 4: 	 from expresion: 'results.get<Obj>(0).get<Int>("value") == 6'
> [2024/04/18 08:38:51.641] 4: 	 with expansion: '3 == 6'
> ```

First error was due to a change in `is_initial_sync_complete` which I now reverted (and rebased this branch in top of master) and it should be gone. I'm investigating the second one.

jbreams commented on src/realm/sync/instruction_applier.cpp:
```diff
@@ -877,87 +877,115 @@ void InstructionApplier::operator()(const Instruction::ArrayErase& instr)
         void on_list(LstBase& list) override
         {
+            // list property
+            REALM_ASSERT(!m_collection_type || *m_collection_type == CollectionType::List);
```
do we want this to be a bad_transaction_log rather than a REALM_ASSERT?

jbreams commented on src/realm/sync/instruction_applier.cpp:
```diff
@@ -877,87 +877,115 @@ void InstructionApplier::operator()(const Instruction::ArrayErase& instr)
         void on_dictionary(Dictionary& dict) override
         {
+            // dictionary property
+            REALM_ASSERT(!m_collection_type || *m_collection_type == CollectionType::Dictionary);
```
Should this be a bad_transaction_log() rather than a REALM_ASSERT?

jbreams commented on src/realm/sync/instruction_applier.cpp:
```diff
@@ -877,87 +877,115 @@ void InstructionApplier::operator()(const Instruction::ArrayErase& instr)
         void on_set(SetBase& set) override
         {
+            // set property
+            REALM_ASSERT(!m_collection_type || *m_collection_type == CollectionType::Set);
```
Should this be a bad_transaction_log() rather than a REALM_ASSERT?

michael-wb reviewed (approved):
LGTM - tests are passing now that the branch has been rebased onto `master`.
